File: packages/core/src/canonical.ts
import type {
  ColumnDefinition,
  DictionaryAttribute,
  DictionaryDefinition,
  MaterializedViewDefinition,
  MaterializedViewRefresh,
  SchemaDefinition,
  SkipIndexDefinition,
  TableDefinition,
  ViewDefinition,
} from './model.js'
import { normalizeKeyColumns } from './key-clause.js'
import { isSchemaDefinition } from './model.js'
import { canonicalizeCodec } from './codec.js'
import { canonicalizeProjection } from './projection.js'
import { normalizeEngine, normalizeSQLFragment } from './sql-normalizer.js'

function sortByName<T extends { name: string }>(items: T[]): T[] {
  return [...items].sort((a, b) => a.name.localeCompare(b.name))
}

function sortKind(kind: SchemaDefinition['kind']): number {
  if (kind === 'table') return 0
  if (kind === 'view') return 1
  if (kind === 'materialized_view') return 2
  return 3
}

function canonicalizeColumn(column: ColumnDefinition): ColumnDefinition {
  return {
    ...column,
    name: column.name.trim(),
    renamedFrom: column.renamedFrom?.trim(),
    type: typeof column.type === 'string' ? column.type.trim() : column.type,
    comment: column.comment?.trim(),
    codec: column.codec ? canonicalizeCodec(column.codec) : undefined,
  }
}

function canonicalizeIndex(index: SkipIndexDefinition): SkipIndexDefinition {
  const expression = normalizeSQLFragment(index.expression)
  if (index.type !== 'text') return { ...index, expression }
  return {
    ...index,
    expression,
    tokenizer: normalizeSQLFragment(index.tokenizer),
    preprocessor: index.preprocessor ? normalizeSQLFragment(index.preprocessor) : undefined,
    postprocessor: index.postprocessor ? normalizeSQLFragment(index.postprocessor) : undefined,
  }
}

function canonicalizeTable(def: TableDefinition): TableDefinition {
  const settings = def.settings
    ? Object.fromEntries(
        Object.entries(def.settings).sort(([a], [b]) => a.localeCompare(b))
      )
    : undefined

  const indexes = def.indexes ? sortByName(def.indexes).map(canonicalizeIndex) : undefined
  const projections = def.projections
    ? sortByName(def.projections).map(canonicalizeProjection)
    : undefined

  const orderBy = normalizeKeyColumns(def.orderBy)
  // ClickHouse derives the primary key from ORDER BY when PRIMARY KEY is
  // omitted. Mirror that so a table with `orderBy` but no `primaryKey` is
  // valid instead of crashing on `undefined.flatMap`.
  const primaryKey = normalizeKeyColumns(def.primaryKey)

  return {
    ...def,
    database: def.database.trim(),
    name: def.name.trim(),
    renamedFrom: def.renamedFrom
      ? {
          database: def.renamedFrom.database?.trim(),
          name: def.renamedFrom.name.trim(),
        }
      : undefined,
    engine: normalizeEngine(def.engine),
    columns: def.columns.map(canonicalizeColumn),
    primaryKey: primaryKey.length > 0 ? primaryKey : orderBy,
    orderBy,
    uniqueKey: def.uniqueKey ? normalizeKeyColumns(def.uniqueKey) : undefined,
    partitionBy: def.partitionBy ? normalizeSQLFragment(def.partitionBy) : undefined,
    ttl: def.ttl ? normalizeSQLFragment(def.ttl) : undefined,
    settings,
    indexes,
    projections,
    comment: def.comment?.trim(),
  }
}

function canonicalizeView(def: ViewDefinition): ViewDefinition {
  return {
    ...def,
    database: def.database.trim(),
    name: def.name.trim(),
    as: normalizeSQLFragment(def.as),
    comment: def.comment?.trim(),
  }
}

const INTERVAL_UNIT_PATTERN = /\b(second|minute|hour|day|week|month|year)s?\b/gi

function canonicalizeInterval(value: string | undefined): string | undefined {
  if (value === undefined) return undefined
  return value
    .trim()
    .replace(/\s+/g, ' ')
    .replace(INTERVAL_UNIT_PATTERN, (unit) => unit.toUpperCase().replace(/S$/, ''))
}

function canonicalizeRefresh(
  refresh: MaterializedViewRefresh | undefined
): MaterializedViewRefresh | undefined {
  if (!refresh) return undefined

  const dependsOn = refresh.dependsOn
    ? [...refresh.dependsOn]
        .map((dep) => ({ database: dep.database.trim(), name: dep.name.trim() }))
        .sort((a, b) => {
          const aKey = `${a.database}.${a.name}`
          const bKey = `${b.database}.${b.name}`
          return aKey.localeCompare(bKey)
        })
    : undefined

  const settings = refresh.settings
    ? Object.fromEntries(
        Object.entries(refresh.settings).sort(([a], [b]) => a.localeCompare(b))
      )
    : undefined

  const canonical: MaterializedViewRefresh = {}
  const every = canonicalizeInterval(refresh.every)
  const after = canonicalizeInterval(refresh.after)
  const offset = canonicalizeInterval(refresh.offset)
  const randomize = canonicalizeInterval(refresh.randomize)
  if (every !== undefined) canonical.every = every
  if (after !== undefined) canonical.after = after
  if (offset !== undefined) canonical.offset = offset
  if (randomize !== undefined) canonical.randomize = randomize
  if (dependsOn && dependsOn.length > 0) canonical.dependsOn = dependsOn
  if (settings && Object.keys(settings).length > 0) canonical.settings = settings
  if (refresh.append) canonical.append = true
  if (refresh.empty) canonical.empty = true
  return canonical
}

function canonicalizeMaterializedView(def: MaterializedViewDefinition): MaterializedViewDefinition {
  const canonical: MaterializedViewDefinition = {
    ...def,
    database: def.database.trim(),
    name: def.name.trim(),
    to: {
      database: def.to.database.trim(),
      name: def.to.name.trim(),
    },
    as: normalizeSQLFragment(def.as),
    comment: def.comment?.trim(),
  }
  const refresh = canonicalizeRefresh(def.refresh)
  if (refresh) {
    canonical.refresh = refresh
  } else {
    delete canonical.refresh
  }
  return canonical
}

function canonicalizeDictionaryAttribute(attribute: DictionaryAttribute): DictionaryAttribute {
  return {
    ...attribute,
    name: attribute.name.trim(),
    type: typeof attribute.type === 'string' ? attribute.type.trim() : attribute.type,
  }
}

function canonicalizeDictionary(def: DictionaryDefinition): DictionaryDefinition {
  const settings = def.settings
    ? Object.fromEntries(Object.entries(def.settings).sort(([a], [b]) => a.localeCompare(b)))
    : undefined

  return {
    ...def,
    database: def.database.trim(),
    name: def.name.trim(),
    renamedFrom: def.renamedFrom
      ? {
          database: def.renamedFrom.database?.trim(),
          name: def.renamedFrom.name.trim(),
        }
      : undefined,
    attributes: def.attributes.map(canonicalizeDictionaryAttribute),
    primaryKey: normalizeKeyColumns(def.primaryKey),
    source: normalizeSQLFragment(def.source),
    layout: normalizeSQLFragment(def.layout),
    lifetime: normalizeSQLFragment(def.lifetime),
    range: def.range
      ? { min: def.range.min.trim(), max: def.range.max.trim() }
      : undefined,
    settings: settings && Object.keys(settings).length > 0 ? settings : undefined,
    comment: def.comment?.trim(),
  }
}

export function canonicalizeDefinition(def: SchemaDefinition): SchemaDefinition {
  if (def.kind === 'table') return canonicalizeTable(def)
  if (def.kind === 'view') return canonicalizeView(def)
  if (def.kind === 'dictionary') return canonicalizeDictionary(def)
  return canonicalizeMaterializedView(def)
}

export function definitionKey(def: SchemaDefinition): string {
  return `${def.kind}:${def.database}.${def.name}`
}

export function canonicalizeDefinitions(definitions: SchemaDefinition[]): SchemaDefinition[] {
  const dedup = new Map<string, SchemaDefinition>()
  for (const def of definitions) {
    const normalized = canonicalizeDefinition(def)
    dedup.set(definitionKey(normalized), normalized)
  }

  return [...dedup.values()].sort((a, b) => {
    const kindOrder = sortKind(a.kind) - sortKind(b.kind)
    if (kindOrder !== 0) return kindOrder
    const dbOrder = a.database.localeCompare(b.database)
    if (dbOrder !== 0) return dbOrder
    return a.name.localeCompare(b.name)
  })
}

export function collectDefinitionsFromModule(mod: Record<string, unknown>): SchemaDefinition[] {
  const out: SchemaDefinition[] = []

  const walk = (value: unknown) => {
    if (!value) return
    if (Array.isArray(value)) {
      for (const entry of value) walk(entry)
      return
    }
    if (isSchemaDefinition(value)) {
      out.push(value)
      return
    }
  }

  for (const value of Object.values(mod)) walk(value)
  return canonicalizeDefinitions(out)
}

File: chkit_python/src/chkit/core/canonical.py
"""Canonicalize definitions for stable comparison and key derivation."""

from __future__ import annotations

import re
from collections.abc import Iterable
from operator import attrgetter
from typing import Final, TypeVar

from chkit.core.codec import canonicalize_codec
from chkit.core.key_clause import normalize_key_columns
from chkit.core.model import (
    ColumnDefinition,
    DictionaryAttribute,
    DictionaryDefinition,
    DictionaryRange,
    MaterializedViewDefinition,
    MaterializedViewRefresh,
    SchemaDefinition,
    SchemaKind,
    SkipIndexDefinition,
    TableDefinition,
    TableRef,
    TableRenamedFrom,
    ViewDefinition,
)
from chkit.core.projection import canonicalize_projection
from chkit.core.sql_normalizer import normalize_engine, normalize_sql_fragment

_T = TypeVar("_T")


def _sort_by_name(items: list[_T]) -> list[_T]:
    # Items always carry a `name: str` attribute (enforced at construction time
    # by the Pydantic models that flow through here).
    return sorted(items, key=attrgetter("name"))


def _sort_kind(kind: SchemaKind) -> int:
    if kind == "table":
        return 0
    if kind == "view":
        return 1
    if kind == "materialized_view":
        return 2
    return 3


def _canonicalize_column(column: ColumnDefinition) -> ColumnDefinition:
    type_value = column.type
    canon_type = type_value.strip() if isinstance(type_value, str) else type_value
    return column.model_copy(
        update={
            "name": column.name.strip(),
            "renamed_from": column.renamed_from.strip()
            if column.renamed_from is not None
            else None,
            "type": canon_type,
            "comment": column.comment.strip() if column.comment is not None else None,
            "codec": canonicalize_codec(column.codec) if column.codec is not None else None,
        }
    )


def _canonicalize_index(index: SkipIndexDefinition) -> SkipIndexDefinition:
    update: dict[str, object] = {"expression": normalize_sql_fragment(index.expression)}
    if index.type == "text":
        update["tokenizer"] = normalize_sql_fragment(index.tokenizer)
        for field in ("preprocessor", "postprocessor"):
            value = getattr(index, field)
            if value is not None:
                update[field] = normalize_sql_fragment(value)
    return index.model_copy(update=update)


def _sorted_settings(
    settings: dict[str, str | int | float | bool] | None,
) -> dict[str, str | int | float | bool] | None:
    if settings is None:
        return None
    return {k: settings[k] for k in sorted(settings.keys())}


def _canonicalize_table(definition: TableDefinition) -> TableDefinition:
    settings = _sorted_settings(definition.settings)
    indexes = (
        [_canonicalize_index(idx) for idx in _sort_by_name(definition.indexes)]
        if definition.indexes is not None
        else None
    )
    projections = (
        [canonicalize_projection(p) for p in _sort_by_name(definition.projections)]
        if definition.projections is not None
        else None
    )
    renamed_from: TableRenamedFrom | None = None
    if definition.renamed_from is not None:
        renamed_from = TableRenamedFrom(
            database=definition.renamed_from.database.strip()
            if definition.renamed_from.database is not None
            else None,
            name=definition.renamed_from.name.strip(),
        )

    normalized_order_by = normalize_key_columns(definition.order_by)
    normalized_primary_key = normalize_key_columns(definition.primary_key)
    # TS canonical.ts: when primary_key is empty, fall back to order_by.
    # Without this, a snapshot written by TS (where omitted PK == order_by)
    # would never match a Python snapshot (where omitted PK == []).
    if not normalized_primary_key:
        normalized_primary_key = list(normalized_order_by)
    return definition.model_copy(
        update={
            "database": definition.database.strip(),
            "name": definition.name.strip(),
            "renamed_from": renamed_from,
            "engine": normalize_engine(definition.engine),
            "columns": [_canonicalize_column(c) for c in definition.columns],
            "primary_key": normalized_primary_key,
            "order_by": normalized_order_by,
            "unique_key": normalize_key_columns(definition.unique_key)
            if definition.unique_key is not None
            else None,
            "partition_by": normalize_sql_fragment(definition.partition_by)
            if definition.partition_by is not None
            else None,
            "ttl": normalize_sql_fragment(definition.ttl) if definition.ttl is not None else None,
            "settings": settings,
            "indexes": indexes,
            "projections": projections,
            "comment": definition.comment.strip() if definition.comment is not None else None,
        }
    )


def _canonicalize_view(definition: ViewDefinition) -> ViewDefinition:
    return definition.model_copy(
        update={
            "database": definition.database.strip(),
            "name": definition.name.strip(),
            "as_": normalize_sql_fragment(definition.as_),
            "comment": definition.comment.strip() if definition.comment is not None else None,
        }
    )


_INTERVAL_UNIT_PATTERN: Final[re.Pattern[str]] = re.compile(
    r"\b(second|minute|hour|day|week|month|year)s?\b", re.IGNORECASE
)


def _canonicalize_interval(value: str | None) -> str | None:
    if value is None:
        return None
    collapsed = re.sub(r"\s+", " ", value).strip()

    def _upper_singular(match: re.Match[str]) -> str:
        token = match.group(0).upper()
        return token.removesuffix("S") if token.endswith("S") else token

    return _INTERVAL_UNIT_PATTERN.sub(_upper_singular, collapsed)


def _canonicalize_refresh(
    refresh: MaterializedViewRefresh | None,
) -> MaterializedViewRefresh | None:
    if refresh is None:
        return None

    depends_on: list[TableRef] | None = None
    if refresh.depends_on is not None:
        depends_on = sorted(
            (
                TableRef(database=dep.database.strip(), name=dep.name.strip())
                for dep in refresh.depends_on
            ),
            key=lambda ref: f"{ref.database}.{ref.name}",
        )

    settings: dict[str, str | int | float] | None = None
    if refresh.settings is not None:
        settings = {k: refresh.settings[k] for k in sorted(refresh.settings.keys())}

    every = _canonicalize_interval(refresh.every)
    after = _canonicalize_interval(refresh.after)
    offset = _canonicalize_interval(refresh.offset)
    randomize = _canonicalize_interval(refresh.randomize)

    payload: dict[str, object] = {}
    if every is not None:
        payload["every"] = every
    if after is not None:
        payload["after"] = after
    if offset is not None:
        payload["offset"] = offset
    if randomize is not None:
        payload["randomize"] = randomize
    if depends_on is not None and len(depends_on) > 0:
        # Use the camelCase alias key for the canonical dict so callers that
        # compare the raw dict to a TS-emitted snapshot see matching keys.
        # ``model_validate`` accepts both forms thanks to
        # ``populate_by_name=True``; ``by_alias=True`` on the inner dump
        # future-proofs us if TableRef ever grows an alias.
        payload["dependsOn"] = [
            d.model_dump(by_alias=True) for d in depends_on
        ]
    if settings is not None and len(settings) > 0:
        payload["settings"] = settings
    if refresh.append:
        payload["append"] = True
    if refresh.empty:
        payload["empty"] = True
    return MaterializedViewRefresh.model_validate(payload)


def _canonicalize_materialized_view(
    definition: MaterializedViewDefinition,
) -> MaterializedViewDefinition:
    canonical_refresh = _canonicalize_refresh(definition.refresh)
    return definition.model_copy(
        update={
            "database": definition.database.strip(),
            "name": definition.name.strip(),
            "to": TableRef(
                database=definition.to.database.strip(),
                name=definition.to.name.strip(),
            ),
            "as_": normalize_sql_fragment(definition.as_),
            "comment": definition.comment.strip() if definition.comment is not None else None,
            "refresh": canonical_refresh,
        }
    )


def _canonicalize_dictionary_attribute(
    attribute: DictionaryAttribute,
) -> DictionaryAttribute:
    type_value = attribute.type
    canon_type = type_value.strip() if isinstance(type_value, str) else type_value
    return attribute.model_copy(
        update={"name": attribute.name.strip(), "type": canon_type}
    )


def _canonicalize_dictionary(definition: DictionaryDefinition) -> DictionaryDefinition:
    settings: dict[str, str | int | float] | None = None
    if definition.settings is not None:
        settings = {k: definition.settings[k] for k in sorted(definition.settings)}
        if len(settings) == 0:
            settings = None

    renamed_from: TableRenamedFrom | None = None
    if definition.renamed_from is not None:
        renamed_from = TableRenamedFrom(
            database=definition.renamed_from.database.strip()
            if definition.renamed_from.database is not None
            else None,
            name=definition.renamed_from.name.strip(),
        )

    return definition.model_copy(
        update={
            "database": definition.database.strip(),
            "name": definition.name.strip(),
            "renamed_from": renamed_from,
            "attributes": [
                _canonicalize_dictionary_attribute(a) for a in definition.attributes
            ],
            "primary_key": normalize_key_columns(definition.primary_key),
            "source": normalize_sql_fragment(definition.source),
            "layout": normalize_sql_fragment(definition.layout),
            "lifetime": normalize_sql_fragment(definition.lifetime),
            "range": DictionaryRange(
                min=definition.range.min.strip(), max=definition.range.max.strip()
            )
            if definition.range is not None
            else None,
            "settings": settings,
            "comment": definition.comment.strip()
            if definition.comment is not None
            else None,
        }
    )


def canonicalize_definition(definition: SchemaDefinition) -> SchemaDefinition:
    if isinstance(definition, TableDefinition):
        return _canonicalize_table(definition)
    if isinstance(definition, ViewDefinition):
        return _canonicalize_view(definition)
    if isinstance(definition, DictionaryDefinition):
        return _canonicalize_dictionary(definition)
    return _canonicalize_materialized_view(definition)


def definition_key(definition: SchemaDefinition) -> str:
    return f"{definition.kind}:{definition.database}.{definition.name}"


def canonicalize_definitions(definitions: Iterable[SchemaDefinition]) -> list[SchemaDefinition]:
    dedup: dict[str, SchemaDefinition] = {}
    for definition in definitions:
        normalized = canonicalize_definition(definition)
        dedup[definition_key(normalized)] = normalized

    return sorted(
        dedup.values(),
        key=lambda d: (_sort_kind(d.kind), d.database, d.name),
    )
